### ISSUE TITLE
Problem about dims in training

### ISSUE THREAD

**Issue by sj-li** [tum-vision/fusenet#1]
When I train the fusenet, I got the following error:

```
F0302 14:54:53.849881  9560 blob.cpp:32] Check failed: shape[i] >= 0 (-1 vs. 0) 
*** Check failure stack trace: ***
    @     0x7f208aedae3d  google::LogMessage::Fail()
    @     0x7f208aedcbc0  google::LogMessage::SendToLog()
    @     0x7f208aedaa23  google::LogMessage::Flush()
    @     0x7f208aedd58e  google::LogMessageFatal::~LogMessageFatal()
    @     0x7f208b4b662b  caffe::Blob<>::Reshape()
    @     0x7f208b578673  caffe::SliceLayer<>::Reshape()
    @     0x7f208b5e8d96  caffe::Net<>::Init()
    @     0x7f208b5ea5e1  caffe::Net<>::Net()
    @     0x7f208b432c0a  caffe::Solver<>::InitTrainNet()
I0302 14:54:53.852614  9567 db_lmdb.hpp:53] shuffle data, start epoch 3
    @     0x7f208b434017  caffe::Solver<>::Init()
    @     0x7f208b4343ba  caffe::Solver<>::Solver()
    @     0x7f208b4a1e83  caffe::Creator_SGDSolver<>()
    @           0x40b978  train()
    @           0x408618  main
    @     0x7f208992d830  (unknown)
    @           0x408d89  _start
Aborted (core dumped)
```
I know the question may be the dimension of net, so I print the size of blob:

```
top size : 3
/home/jason/Jason/Code/Semantic_Segmentation/fusenet/src/caffe/layers/slice_layer.cpp, 56
2, 3, 480, 640
/home/jason/Jason/Code/Semantic_Segmentation/fusenet/src/caffe/layers/slice_layer.cpp, 56
2, 1, 480, 640
/home/jason/Jason/Code/Semantic_Segmentation/fusenet/src/caffe/layers/slice_layer.cpp, 56
2, -1, 480, 640
```

So I think maybe I organize the data in a wrong way. I use the following list file to make lmdb dataset by order color images, depth images, label images, use the default convert_imagedataset offered by caffe:

```
00002_colors.png
00003_colors.png
...
00002_depth.png
00003_depth.png
...
00002_label.png
00003_label.png
```

And I wonder both label image and depth are one channel, how can the net classify them?

At the mean time , I am not clearly know how to use weighted cross-entropy, can you help me?

Thanks all the same.

**Comment by hazirbas** [comment]
Hi,
Can you please use the provided script to write lmdb files? Network expect 4-dimensional (RGB-D) input which RGB and Depth will be sliced in the network. Depth might be 1-D float or unsigned char but label images must be 1-channel unsigned char.

Weighted cross entropy defined in [1]. Please refer to the paper for details. 
`w(c) = median_freq / freq(c),`
where freq(c) is the number of pixels of class c divided by the total number of pixels in images where c is present, and median freq is the median of these frequencies.

[1] - Eigen, D., Fergus, R.: Predicting depth, surface normals and semantic labels with
a common multi-scale convolutional architecture. ICCV (2015)

**Comment by sj-li** [comment]
Thanks for you reply!

Well, I don't find the scripts to write lmbd file( demo   ./fusenet/scripts/save_lmdb.py  list in the readme ), So i use the caffe default program convert_imageset to produce lmbd data(see https://github.com/BVLC/caffe/blob/master/tools/convert_imageset.cpp). I keep the rgb 3 channel, depth 1 chanel(range from 0~1), and ground truth 1channel(i use 13 label)， and put all these files in a txt file like above(May be I should group the corresponding term in the same line and split them by space, not just put one image for one line？)

**Comment by hazirbas** [comment]
Hi.
As far as I remember, default caffe lmdb writer does not accept 4-D input image. So you have to write your own program to create lmdb dataset.

**Comment by sj-li** [comment]
Well, thanks to your suggestion!
I directly a layer that read images from file and pass them to the net, and it runs.
However, the loss and accuracy almost not change from the begin!
begin:
```
I0309 09:12:32.442052  2945 solver.cpp:238] Iteration 0, loss = 3.53627
I0309 09:12:32.442072  2945 solver.cpp:254]     Train net output #0: accuracy = 0.0807686
I0309 09:12:32.442080  2945 solver.cpp:254]     Train net output #1: loss = 3.53627 (* 1 = 3.53627 loss)
I0309 09:12:32.442088  2945 sgd_solver.cpp:106] Iteration 0, lr = 1e-10
I0309 09:12:41.088881  2945 solver.cpp:238] Iteration 20, loss = 3.68417
I0309 09:12:41.088903  2945 solver.cpp:254]     Train net output #0: accuracy = 0.0847367
I0309 09:12:41.088910  2945 solver.cpp:254]     Train net output #1: loss = 3.48979 (* 1 = 3.48979 loss)
I0309 09:12:41.088914  2945 sgd_solver.cpp:106] Iteration 20, lr = 1e-10
I0309 09:12:49.948145  2945 solver.cpp:238] Iteration 40, loss = 3.64031
I0309 09:12:49.948168  2945 solver.cpp:254]     Train net output #0: accuracy = 0.0625678
I0309 09:12:49.948175  2945 solver.cpp:254]     Train net output #1: loss = 3.71061 (* 1 = 3.71061 loss)

```
now:

```
I0309 10:19:03.482935  2945 sgd_solver.cpp:106] Iteration 9160, lr = 1e-10
I0309 10:19:12.166920  2945 solver.cpp:238] Iteration 9180, loss = 3.63868
I0309 10:19:12.166957  2945 solver.cpp:254]     Train net output #0: accuracy = 0.0753186
I0309 10:19:12.166965  2945 solver.cpp:254]     Train net output #1: loss = 3.61206 (* 1 = 3.61206 loss)
I0309 10:19:12.166970  2945 sgd_solver.cpp:106] Iteration 9180, lr = 1e-10
I0309 10:19:20.847422  2945 solver.cpp:238] Iteration 9200, loss = 3.62983
I0309 10:19:20.847520  2945 solver.cpp:254]     Train net output #0: accuracy = 0.0885214
I0309 10:19:20.847538  2945 solver.cpp:254]     Train net output #1: loss = 3.52476 (* 1 = 3.52476 loss)
I0309 10:19:20.847543  2945 sgd_solver.cpp:106] Iteration 9200, lr = 1e-10
I0309 10:19:29.531528  2945 solver.cpp:238] Iteration 9220, loss = 3.65184
I0309 10:19:29.531558  2945 solver.cpp:254]     Train net output #0: accuracy = 0.0773005
I0309 10:19:29.531569  2945 solver.cpp:254]     Train net output #1: loss = 3.57626 (* 1 = 3.57626 loss)
I0309 10:19:29.531574  2945 sgd_solver.cpp:106] Iteration 9220, lr = 1e-10
```

Is iterate time too small? Or some wrong must happen?

**Comment by hazirbas** [comment]
I see that you set the learning rate to 1e-10 which is a lot smaller than what we have used in the paper (0.001). Moreover, do not forget that default accuracy layer in caffe does not compute the segmentation accuracy, it is rather implemented for classification.

**Comment by sj-li** [comment]
Well, Thanks very much, it's the first time that I train a net, so face many problem ):

 And now when I test my net by my own script:

```
import caffe

import numpy as np
import os
import os.path
from skimage.io import imsave, imread, imshow

weights_ = '/home/jason/Jason/Code/Semantic_Segmentation/fusenet/fusenet/segmentation/nyu40-sf1/snapshot_iter_64000.caffemodel'
deploy_ = '/home/jason/Jason/Code/Semantic_Segmentation/fusenet/fusenet/segmentation/nyu40-sf1/deploy.prototxt'
depth_list_ = '/home/jason/Jason/Code/Semantic_Segmentation/fusenet/fusenet/segmentation/nyu40-sf1/test/list/depth.txt'
rgb_list_ = '/home/jason/Jason/Code/Semantic_Segmentation/fusenet/fusenet/segmentation/nyu40-sf1/test/list/rgb.txt'
save_path = '/home/jason/Jason/Code/Semantic_Segmentation/fusenet/fusenet/segmentation/nyu40-sf1/test/result'

fusenet = caffe.Net(deploy_, weights_, caffe.TEST)

caffe.set_device(0)
caffe.set_mode_gpu()

depth_list = np.loadtxt(depth_list_, dtype=str)
rgb_list = np.loadtxt(rgb_list_, dtype=str)

for file in zip(rgb_list, depth_list):

    rgb = imread(file[0])
    fusenet.blobs['rgb'].reshape(rgb.shape)
    fusenet.blobs['rgb'].data[...] = rgb

    depth = imread(file[1])
    fusenet.blobs['depth'].reshape(rgb.shape)
    fusenet.blobs['depth'].data[...] = rgb

    fusenet.forward()
    out = fusenet.blobs['argmax'].data
    imsave(os.path.join(save_path, os.path.basename(file[1])), out)
```

It is strange that the script can not find the "Input" layer(when I comment the Input, it can not find unpooling layer, I think the problem maybe the same). and I check the compile of caffe and confirm that input_layer is compiled successfully:

```
[libprotobuf ERROR google/protobuf/text_format.cc:274] Error parsing text-format caffe.NetParameter: 6:15: Message type "caffe.LayerParameter" has no field named "input_param".
WARNING: Logging before InitGoogleLogging() is written to STDERR
F0311 02:58:09.698727 23120 upgrade_proto.cpp:68] Check failed: ReadProtoFromTextFile(param_file, param) Failed to parse NetParameter file: /home/jason/Jason/Code/Semantic_Segmentation/fusenet/fusenet/segmentation/nyu40-sf1/deploy.prototxt
*** Check failure stack trace: ***
[1]    23120 abort (core dumped)  python test.py
```

Thank you very much for you answer so many question :)

**Comment by hazirbas** [comment]
Which caffe version are you using? Can you try the older syntax:

input: "rgb"
input_shape {
  dim: 1
  dim: 3
  dim: Height
  dim: Width
}

**Comment by sj-li** [comment]
Well, I just use the caffe in fusenet.
And I try two version:
First:
```
layer {
  name: "rgb"
  type: "Input"
  top: "rgb"
  input_param { shape {dim: 1 dim: 3 dim: 240 dim: 320} }
}
layer {
  name: "depth"
  type: "Input"
  top: "depth"
  input_param { shape {dim: 1 dim: 1 dim: 240 dim: 320} }
}
```

error is:

```
[libprotobuf ERROR google/protobuf/text_format.cc:274] Error parsing text-format caffe.NetParameter: 6:15: Message type "caffe.LayerParameter" has no field named "input_param".
WARNING: Logging before InitGoogleLogging() is written to STDERR
F0311 03:54:54.355470 24962 upgrade_proto.cpp:68] Check failed: ReadProtoFromTextFile(param_file, param) Failed to parse NetParameter file: /home/jason/Jason/Code/Semantic_Segmentation/fusenet/fusenet/segmentation/nyu40-sf1/deploy.prototxt
*** Check failure stack trace: ***
[1]    24962 abort (core dumped)  python test.py
```
Second:

```
layer {
input: "rgb"
input_shape {
dim: 1
dim: 3
dim: Height
dim: Width
}
}
layer {
input: "depth"
input_shape {
dim: 1
dim: 1
dim: Height
dim: Width
}
}
```

error is 

```
[libprotobuf ERROR google/protobuf/text_format.cc:274] Error parsing text-format caffe.NetParameter: 3:6: Message type "caffe.LayerParameter" has no field named "input".
WARNING: Logging before InitGoogleLogging() is written to STDERR
F0311 03:56:45.960330 25143 upgrade_proto.cpp:68] Check failed: ReadProtoFromTextFile(param_file, param) Failed to parse NetParameter file: /home/jason/Jason/Code/Semantic_Segmentation/fusenet/fusenet/segmentation/nyu40-sf1/deploy.prototxt
*** Check failure stack trace: ***
[1]    25143 abort (core dumped)  python test.py

```

The strange thing is C++ version of layer is ok, but use python it goes wrong. Do I need do some thing when want to use layer in python?

**Comment by hazirbas** [comment]
I am afraid, you have to consult this on the official caffe repository.

**Comment by sj-li** [comment]
Well, you are right, it is the version problem, I use the 

```
input: "rgb"
input_shape {
dim: 1
dim: 3
dim: 240
dim: 320
}

input: "depth"
input_shape {
dim: 1
dim: 3
dim: 240
dim: 320
}
```

it works, and now it can not find the unpooling layer:

```
[libprotobuf ERROR google/protobuf/text_format.cc:274] Error parsing text-format caffe.NetParameter: 1049:19: Message type "caffe.LayerParameter" has no field named "unpooling_param".
WARNING: Logging before InitGoogleLogging() is written to STDERR
F0311 04:20:39.641866 26405 upgrade_proto.cpp:68] Check failed: ReadProtoFromTextFile(param_file, param) Failed to parse NetParameter file: /home/jason/Jason/Code/Semantic_Segmentation/fusenet/fusenet/segmentation/nyu40-sf1/deploy.prototxt
*** Check failure stack trace: ***
[1]    26405 abort (core dumped)  python test.py
```

But the syntax is the same in train.prototxt and deploy.prototxt, and I also check the caffe.proto and find the correspondings:

```
layer {
  name: "unpool5"
  type: "Unpooling"
  bottom: "pool5"
  bottom: "pool5_mask"
  top: "pool5_D"
  unpooling_param {
    output_h: 15
    output_w: 20
  }
}
```

Is there some thing wrong? I find maybe that there is not an official version of unpooling_layer in caffe

**Comment by hazirbas** [comment]
Please check your caffe pythonpath settings in your bashrc. You may link to a newer version of caffe which doesnt include our caffe python probably.

**Comment by sj-li** [comment]
Well, all things go well! You are so kind, thank you very for help me, I think maybe I face all problems a new learner will face :D
